src/librustc_mir/interpret/operator.rs

```diff
@@ -60,32 +60,102 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
         let left = left.to_scalar()?;
         let right = right.to_scalar()?;
 
-        let left_kind = match left_layout.abi {
-            layout::Abi::Scalar(ref scalar) => scalar.value,
-            _ => return err!(TypeNotPrimitive(left_layout.ty)),
-        };
-        let right_kind = match right_layout.abi {
-            layout::Abi::Scalar(ref scalar) => scalar.value,
-            _ => return err!(TypeNotPrimitive(right_layout.ty)),
-        };
         trace!("Running binary op {:?}: {:?} ({:?}), {:?} ({:?})",
-        bin_op, left, left_kind, right, right_kind);
+            bin_op, left, left_layout.ty.sty, right, right_layout.ty.sty);
 
-        // I: Handle operations that support pointers
-        if !left_kind.is_float() && !right_kind.is_float() {
-            if let Some(handled) =
-                M::try_ptr_op(self, bin_op, left, left_layout, right, right_layout)?
-            {
-                return Ok(handled);
+        // Handle non-integer operations
+        if let ty::Char = left_layout.ty.sty {
```

Comment:
can you make these `if let`s a match over the type + method calls instead of having the logic right here?

Reply:
You want one function per type?

Reply:
if that's not too much hazzle, yes

Reply:
Done (with next push)